Add wet run testing to github

mackenziesnyder opened the pull request:
MSE results:
- run 1: 1.0965
- run 2: 1.0965

To do:
- decide on acceptable MSE, right now I've set it to 1.2mm (all 32 afids, x, y, z included). 

Note: 
- wet runs take around 20 minutes to run due to model downloads 
- just for python3.10 

for issue: #41

Dhananjhay commented:
Superb job @mackenziesnyder! I was wondering if it's possible to cache the CNN models so that we don't have to download them everytime a wet-run test is executed?

mackenziesnyder commented:
> Superb job @mackenziesnyder! I was wondering if it's possible to cache the CNN models so that we don't have to download them everytime a wet-run test is executed?

i will look into this! that would make the run a lot faster

mackenziesnyder commented:
> We can reproduce the MSE to the exact decimal, i.e., 1.0965 on every run, so I was wondering if we need a threshold to begin with. What do you guys think? @mackenziesnyder @ataha24. Wet-run testing is basically to run the pipeline and make sure the rules are running properly and we can get a .fscv file out in the end.

I can set it to this exact value!

ataha24 commented:
Sounds good to me!

mackenziesnyder commented:
@Dhananjhay @ataha24 I've made the wet runs to be triggered on pull request, added --fidqc and --stereotaxy STN to the wet run, and removed the code that had attempted to cache the model dir. This one should be good to go!

Dhananjhay commented:
This looks awesome, @mackenziesnyder! One small thing - I just merged PR #46 where I removed the logic of running dry run tests using `poethepoet` and I was wondering if we could achieve the same here. You can refer to the main branch for guidance. Thanks again for the amazing work!

mackenziesnyder commented:
> This looks awesome, @mackenziesnyder! One small thing - I just merged PR #46 where I removed the logic of running dry run tests using `poethepoet` and I was wondering if we could achieve the same here. You can refer to the main branch for guidance. Thanks again for the amazing work!

Will do!! I will refactor it to match !

Dhananjhay commented:
@mackenziesnyder I was inspecting the wet-run test, is there a way to reduce the verbosity of the rule which is responsible for downloading the ‘model.zip’ file? Right now the terminal gets flooded with the downloading logs/progress.

mackenziesnyder commented:
> @mackenziesnyder I was inspecting the wet-run test, is there a way to reduce the verbosity of the rule which is responsible for downloading the ‘model.zip’ file? Right now the terminal gets flooded with the downloading logs/progress.

Yes I will look into this!

mackenziesnyder commented:
> > Yes I will look into this, i was thinking we could add a log file to the model download rule if we also want to clear up the terminal for users too, what do you think @Dhananjhay ?
> 
> I don't recall the terminal being flooded with the download rule when we were running autoafids locally on our servers. CBS is still not working so we might have to wait to make that decision. But it's a good idea regardless, something to think about.

Ok, I added the -q to the model download, which will only show output if something fails. I think it could be a feasible solution for now!

mackenziesnyder commented:
@Dhananjhay I separated the dry runs and wet run into different files so they could run on different github functions (push vs. pull request) please let me know if you think this implementation is ok - if so I believe this pr is good to merge

Dhananjhay commented:
Perfect! We might have to hold off until PR #45 is merged because it'd change the way we run these wet tests, but only so slightly, i.e., we will conda on the runner and run the same autoafids command that we are currently running but with an extra flag `--use-conda`. I'm waiting for the CBS server to become functional again before I can merge that PR. I'll keep you posted, great work!

mackenziesnyder commented:
> Perfect! We might have to hold off until PR #45 is merged because it'd change the way we run these wet tests, but only so slightly, i.e., we will conda on the runner and run the same autoafids command that we are currently running but with an extra flag `--use-conda`. I'm waiting for the CBS server to become functional again before I can merge that PR. I'll keep you posted, great work!

Sounds good!!

mackenziesnyder commented:
@Dhananjhay I have tested this out and it ran successfully, I had to add an additional flag `--conda-frontend conda` because of a mamba dependency with conda. I think this is specific to the fact we do the dry and wet runs with poetry

Dhananjhay commented:
Amazing work @mackenziesnyder! I'll review this PR later today and merge it in, thank you!

mackenziesnyder commented:
> Amazing work @mackenziesnyder! I'll review this PR later today and merge it in, thank you!

thanks!

Dhananjhay commented:
- Replaced the Poetry-based environment with a Conda-based environment (run the pipeline by calling the run.py executable).
- Implemented caching for Conda environments to reduce setup time.
- Cached CNN models to avoid repeated downloads.
- Resulted in optimizing the wet run testing timeline by ~ 58%.
- Fixed some dry run tests by defining number of cores, i.e., `c1` flag.

mackenziesnyder commented on .github/workflows/wetrun-testing.yml:
```diff
@@ -0,0 +1,59 @@
+        run: |
+          conda activate snakebids-env
+          ./autoafids/run.py test_data/bids_wetrun_testing test_out participant \
+            --cores all --force-output --stereotaxy STN --fidqc --use-conda --conda-frontend mamba | tee autoafids_output.log
```
do we still need the --conda-frontend mamba here? I added this because it was needed when originally using poetry but when I activated a conda env and ran autoafids during a test on cbs I no longer needed this flag

Dhananjhay replied:
Yeah, it's a little tricky to run with poetry but mainly we use it here to speed up the pipeline. When using conda, Snakemake by default injects `--conda-frontend conda` to resolve the dependencies; we instead use  `--conda-frontend mamba` which is much faster at that!
